kmip/services/server/engine.py

```diff
@@ -0,0 +1,404 @@
+from kmip.core.messages.payloads import query
+
+from kmip.core import misc
+
+from kmip.services.server.crypto import engine
+
+
+class KmipEngine(object):
+    """
+    A KMIP request processor that acts as the core of the KmipServer.
+
+    The KmipEngine contains the core application logic for the KmipServer.
+    It processes all KMIP requests and maintains consisten state across
+    client connections.
+
+    Features that are not supported:
+        * KMIP versions > 1.1
+        * Numerous operations, objects, and attributes.
+        * User authentication
+        * Batch processing options: UNDO
+        * Asynchronous operations
+    """
+
+    def __init__(self):
+        """
+        Create a KmipEngine.
+        """
+        self._logger = logging.getLogger(__name__)
+
+        self._cryptography_engine = engine.CryptographyEngine()
+        self._lock = threading.RLock()
+
+        self._id_placeholder = None
+
+        self._protocol_versions = [
+            contents.ProtocolVersion.create(1, 1),
+            contents.ProtocolVersion.create(1, 0)
+        ]
+
+        self._protocol_version = self._protocol_versions[0]
+
+    def _kmip_version_supported(supported):
+        def decorator(function):
+            def wrapper(self, *args, **kwargs):
+                if float(str(self._protocol_version)) < float(supported):
+                    name = function.__name__
+                    operation = ''.join(
+                        [x.capitalize() for x in name[9:].split('_')]
+                    )
+                    raise exceptions.OperationNotSupported(
+                        "{0} is not supported by KMIP {1}".format(
+                            operation,
+                            self._protocol_version
+                        )
+                    )
+                else:
+                    return function(self, *args, **kwargs)
+            return wrapper
+        return decorator
+
+    def _synchronize(function):
+        def decorator(self, *args, **kwargs):
+            with self._lock:
+                return function(self, *args, **kwargs)
+        return decorator
+
+    def _set_protocol_version(self, protocol_version):
+        if protocol_version in self._protocol_versions:
+            self._protocol_version = protocol_version
+        else:
+            raise exceptions.InvalidMessage(
+                "KMIP {0} is not supported by the server.".format(
+                    protocol_version
+                )
+            )
+
+    def _verify_credential(self, request_credential, connection_credential):
+        # TODO (peterhamilton) Add authentication support
+        # 1. If present, verify user ID of connection_credential is valid user.
+        # 2. If present, verify request_credential is valid credential.
+        # 3. If both present, verify that they are compliant with each other.
+        # 4. If neither present, set server to only allow Query operations.
+        pass
+
+    @_synchronize
```

Comment:
Yes. For now, requests are handled one at a time, one per connection thread. In reality, this doesn't have to be the case. While a single `KmipEngine` will be shared across threads, synchronization only really has to protect data store operations (e.g., data creation, data persistence, data management). Multiple operations could be in execution simultaneously under this model. Some of the state-based features (e.g., the ID placeholder, the protocol version per request) could be persisted to the data store as well.

The reactor pattern could certainly work here. There are many options. I went with the thread-per-connection approach because it was the solution I knew I could get working fastest, since I've written this type of system before. To manage overhead, I'm capping the number of simultaneous connections to 5 (which is the recommended cap for `socket.listen`).

I had actually looked at Twisted several months ago when looking at Python server libraries. While powerful, at the time it looked like I would've had to figure out how to couple `kmip.core` encoding/decoding to a Twisted `Protocol` class, in addition to teasing out the SSL/TLS micromanagement that KMIP requires. It seemed like it would be hard to get working right. I'm sure I could've figured it out, but that was my thinking at the time.